### DISCUSSION TITLE
Denoise audio samples with 'afftdn'

### DISCUSSION THREAD

**Discussion by 33kelder** [ffmpeginteropx/FFmpegInteropX#256] [Q&A] [answered]
I came across FFmpegInteropX because it contains a FrameGrabber class which is significantly faster than the MediaComposition.GetThumbnailAsync. Curious of what more functionality FFmpeg has to offer, I also saw the functionality of denoising audio samples using the filter 'afftdn'. But how would I be able to use this in an UWP app using the FFmpegInteropX nuget package?
Ronald

**Comment by lukasf** [comment, marked as answer]
Hi @33kelder, yes you can use almost any ffmpeg filter if you use FFmpegInteropX for playback. For audio filters, call e.g. `SetFFmpegAudioFilters("afftdn=nr=10:nf=-40")`. You can set this anytime during playback. Filters can be chained using `,` as separator.

In our sample apps, you can play with this: Open a file, open the panel on the left, scroll down to the "FFmpeg effects" section, enter or update the filter string and press "Return" key to apply.

Check ffmpeg docs for details on filter usage:
https://ffmpeg.org/ffmpeg-filters.html

**Reply by 33kelder** [reply]
Hi @lukasf , thank you for your quick answer. I realize I wasn't clear enough about my scenario, but my use case is not for playback. It is for processing files, so file in and file out. All regular FFmpeg samples/docs I find are about processing files. Is that also possible for FFmpegInteropX in an UWP app? And if so, do you know of some sample which could get me started?

**Reply by lukasf** [reply]
Our library is currently a pure decoding library, intended for playback scenarios. I never tried this myself, but you should be able to use our MediaStreamSource with the UWP transcoding APIs to generate file output. Check the `MediaTranscoder` class for that. If you use audio effects in our library, you should see the effect in the generated file from MediaTranscoder. But using MediaTranscoder means that for the encoding part itself, you are limited to UWP standards, so you cannot freely choose and configure codecs like it would be possible in ffmpeg.

**Reply by 33kelder** [reply]
Thanks, I will give it a try.

**Reply by 33kelder** [reply]
I tried it with the following code but I always get a null reference exception when the TranscodeAsync method is called. I played around with several parameter like BufferTime, IsLive, AlwaysReencode, and HardwareAccelerationEnabled, but I couldn't find a configuration that works. So I guess I doesn't work at all unfortunately.
`            
var mediaStreamSource = ffmpeg.GetMediaStreamSource();
            mediaStreamSource.BufferTime = TimeSpan.Zero;
            mediaStreamSource.IsLive = false;
            using (var outputStream = await audioFile.OpenAsync(FileAccessMode.Read))
            {
                var mediaTranscoder = new MediaTranscoder()
                {
                    AlwaysReencode = true,
                    HardwareAccelerationEnabled = true,
                };
                PrepareTranscodeResult mediaEncodingProfile = MediaEncodingProfile.CreateMp3(AudioEncodingQuality.Low);
                prepareTranscodeResult = await mediaTranscoder.PrepareMediaStreamSourceTranscodeAsync(
                    mediaStreamSource,
                    outputStream,
                    mediaEncodingProfile
                    );
                if (prepareTranscodeResult.CanTranscode)
                {
                    await prepareTranscodeResult.TranscodeAsync();
                }
            }
`

**Reply by brabebhin** [reply]
Your output stream is opened for read, shouldn't it be open for write?
In ffmpeg media source config, you should probably be disabling all traces of hardware acceleration to prevent compressed samples.

**Reply by 33kelder** [reply]
Yes, that was the problem. Thank you for having a look at this! This opens a world of possibilities for me.

**Reply by brabebhin** [reply]
Yes. Theoretically we could extend our library to do encoding as well,  but we never had the time to do it and always were playback related ideas that took priority. If you want to contribute to the project,  feel free to open PRs.
